File: 108-convert-sorted-array-to-binary-search-tree/108-convert-sorted-array-to-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def sortedArrayToBST(self, nums):
        """
        :type nums: List[int]
        :rtype: TreeNode
        """
        return self.constructBST(nums, 0, len(nums)-1)
    
    def constructBST(self, nums, start, end):
        if end < start: return None
        middle = (start + end) // 2
        root = TreeNode(nums[middle])
        root.left = self.constructBST(nums, start, middle -1)
        root.right = self.constructBST(nums, middle + 1, end)
        return root
        

File: 108-convert-sorted-array-to-binary-search-tree/108-convert-sorted-array-to-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode sortedArrayToBST(int[] nums) {
        return constructBST(nums, 0, nums.length-1);
    }
    
    public TreeNode constructBST(int[] nums, int start, int end) {
        if (end < start) return null;
        int middle = (start + end) / 2;
        TreeNode root = new TreeNode(nums[middle]);
        root.left = constructBST(nums, start, middle - 1);
        root.right = constructBST(nums, middle + 1, end);
        return root;
    }
}